### ISSUE TITLE
Harmonize action/record/selectors query location in function argument

### ISSUE THREAD

**Issue by ssweber** [PySimpleSQL/pysimplesql#103]
Records are setup like:
`ss.record('Journal.mood_id', sg.Combo…`

Selectors and Actions put the query name second:
```
[ss.selector('sel_journal','Journal',sg.Table,num_rows=10,headings=headings)],
     [ss.actions('act_journal','Journal',edit_protect=False)],
```

unless backwards compatibility is a major concern, it might make sense to flip and put the form/query name first in selector/actions. And make the key a named argument.

Eg
`[ss.selector('Journal',key=“sel_journal”,sg.Table,num_rows=10,headings=headings)],`

**Comment by ssweber** [comment]
Also, is it even necessary to have the user define a key for actions/records?

They could still put them in if they wanted, or we could say: if none given, it will be 
query_selector_1 (2/3/….)
query_action.insert_1 (1/2/3…)

Then selectors could be same as records

[ss.selector(“Journal”,sg.Table,key=[Optional], num_rows=10,headings=headings)],

**Comment by ssweber** [comment]
When I first started using the library, I didn’t know at first glance if sel_journal and act_journal was special, like in records with query.column 🤣

**Comment by PySimpleSQL** [comment]
I'll leave this open - it's not a bad idea.  This will be a v3.0 release at this point with all of the changes, so breaking reverse compatibility isn't a huge deal.  

As far as the selectors needing a key,  the key is how they are currently found (when updating selectors for example), as the map @ Query.selector needs to map a selector against a known element.  I'll have to revisit the code, but this was one of the reasons it needed a unique key, another being that process_events needs to look for selector events by the element key too.

In an early version, selectors were more similar to records - 'Journal.SELECTOR' as the name would make it a selector, but i ended up finding it a little confusing. I'm open to ideas to make it more user friendly, we just need some method to be able to detect when an element is a selector to be able to act on it. If just switching those parameters around accomplishes that, then that may be the best way forward

**Comment by ssweber** [comment]
You now use metadata to designate it a selector: `type': TYPE_SELECTOR`

So I don’t believe we’ll need user to provide key. I’ll test.

with the new header.add(). I think if the query/column is always first for all the convience functions, that’ll make explaining the arguments easy.

**Comment by PySimpleSQL** [comment]
It goes pretty far beyond just the metada - the metadata helps sort things
out at binding in auto_map_elements, but it's also used in
Query.selectors[] (the actual element<->Query map), update_elements,
process_events, and a couple more places.  Plus I think the user would want
to know the key of their selector in case they want to watch for events and
add custom functionality, or do callbacks - in fact, I do this in a few
different projects.

**Comment by ssweber** [comment]
#104 is a quick example, with a modified journal_internal. 

Also, changing the timeout to 1 stopped that _TIMEOUT_ event from firing.

**Comment by ssweber** [comment]
I’m sorry, that is a messy pull request. I’m going to close, cleanup and do it again.

**Comment by ssweber** [comment]
That pull request wouldn’t merge cleanly after all your progress… what I had proposed was:

```
def actions(query, key=None…
key=f'action_{table}' if key is None else key

def selector(table, element=sg.LBox, key=None…
key=f'selector_{table}' if key is None else key

class TableHeadings
def add_column(self, column_name:str, heading_column:str…
```

**Comment by ssweber** [comment]
Never mind 😄 

I see in your edits, ‘table’ in def record() is now key, and have it using it instead of a user supplied key.

I’m going to stop obsessing about this and work on the Message Pack next

**Comment by PySimpleSQL** [comment]
Yeah, I'm not sure how or when it happened, but both table and key ended up
being one and the same later in the code, so one of them was redundant.  I
found a few oddities like that last night

**Comment by ssweber** [comment]
Merging key with table name gave me another idea. 

Last idea, I promise! #112

**Comment by PySimpleSQL** [comment]
Thanks, I will look at this next.  Going to get sorting working on fk
relationship description columns instead of fk first

**Comment by PySimpleSQL** [comment]
I was just reviewing some code - this might be a little deeper than just changing some naming conventions.  For example, save_record(), record() and auto_map_elements() all use the element key - so a new way of transferring this information might be needed.  For example, during auto_map_elements, the table name and column names are extracted from the key so that it knows which query to map the element to.  If we use : as a way of supplying an alternate key, then the table and column information are lost by the time it gets to auto_map_elements().  The same goes for save_record().  I guess it could be added to metadata instead - though I try to release the metadata back to the user after everything is created so that it is just a temporary shuttle for this information.  This means that element_map may need extended to hold this information.

I'm going to have to think on this some more to see if all of this will be worth the effort just for the sake of selector key naming more similar to record key naming.

**Comment by ssweber** [comment]
Sorry, I’m not explaining well enough.

It’s not passing an alternative key, it’s in addition to the table.column. So very little code has to change. It only needs to split at ‘:’ so as not to grab the unique identifier supplied by the user when determining table and column.

What I’ll do is finish the implementation, and then fix up all the examples… and you can take a look at before v after to decide.

**Comment by PySimpleSQL** [comment]
Ok, that would be helpful.  Thanks!

**Comment by PySimpleSQL** [comment]
I haven't forgot about this.  A lot of changes the last day or two paved the way to make this kind of switch easier, especially with the ElementMap cleaned up and holding the data we need.

At minimum, ss.actions() and ss.selector() should have the table name first for sure.  I will probably start there and then see if it makes sense to chase any further

**Comment by ssweber** [comment]
Sorry, still chipping away at checkboxes. Almost have them working, regardless of how they are entered...

My idea for this was to make the structure like:
action('table.column?where', element=, key=None...

Parse the table, column, where, and put them in metadata for auto-mapping. Change auto-mapping to look for metadata, instead of parsing the key.
If key is provided, it now can be whatever the user wants.
If none is provided, just use the 'table.column?where'


selector('table', element=, key=None...

If key is provided, use it
If none is provided, just use 'table'

actions('table', key=None')

if key is provided, use it
if none is provided, just use 'table'+.insert_record, ...

**Comment by ssweber** [comment]
Yes, and your work in ElementMap makes this really easy now. I think making key a named argument, past the element=, makes sense, because PySimpleGui doesn't require keys, but users are used to adding them later if they need them for event-watching.

So initially, if someone doesn't want to add keys, they can just focus on creating the layout

**Comment by PySimpleSQL** [comment]
I did the refactor.  I really like it - it cleans up the look of the layout creation a ton.

Ok, I see more now what you mean.  I will work on this next.
So the plan is:
ss.record('Journal.title', key='mykey') # Journal.title is the key unless specified by the optional  key=
ss.selector('Journal', key='mytable') # key will be 'selector' (through keygen) if no key is supplied, otherwise the supplied optional key
ss.actions('Journal', key='myaction') # key will be Journal.insert_record.... unless a different key is passed in optionally

**Comment by ssweber** [comment]
To be clear, the idea was:

```
def actions(table_name:str, key=None # Journal.title is the key unless specified by the optional key=
def selector(table_name: str, element: sg.Element = sg.LBox, key=None ... # key will be 'Journal' (through keygen) if no key is supplied, otherwise the supplied optional key
def record(table.column?where=, element:sg.Element=sg.I, key=None ... key will be Journal.insert_record.... unless a different key is passed in optionally

**Comment by ssweber** [comment]
That way, if user doesn't want to supply keys, it could just be:
```
    selector = [ss.selector('person',sg.Table, num_rows=4,headings=headings,auto_size_columns=True)]
    ss.record("person.example", sg.Checkbox)],
    [ss.actions("person", default=True)],
```

or they could supply keys:

```
    selector = [ss.selector('person',sg.Table, key='myuniquetablekey', num_rows=4,headings=headings,auto_size_columns=True)]
    ss.record("person.example", sg.Checkbox, key='recordkey')],
    [ss.actions("person", key='personaction', default=True)],

```

**Comment by ssweber** [comment]
You're so fast! trying to catch you before its done, lol

**Comment by PySimpleSQL** [comment]
This was a big one.  But it's soooo nice!

I varied just a little from what we talked about, but the end result is great
Here is the new character meanings: '.'=table/column separator, ':' = additional information, '!'= keygen separtor

Selectors get the name key:selector, where key defaults to the table name (Journal:selector)
Actions get the name key:action, where key defaults to the table name (Journal:table_next)
records get the key name, where key defaults to the record name (Journal.mood_id)

The required marker follows this too.  Journal.mood_id:marker

And for the keygen, '!' itself hints to an issue (a key collision), so reading the following makes perfect sense to me:
Journal:selector, Journal:selector!1, Journal:selector!2, etc...

Here is the window keys dict from the journal_internal example, where no keys were specified:
dict_keys(['Journal:selector', 'Journal:db_save', 'Journal:table_first', 'Journal:table_previous', 'Journal:table_next', 'Journal:table_last', 'Journal:table_duplicate', 'Journal:table_insert', 'Journal:table_delete', 'Journal:search_input', 'Journal:search_button', 'Journal.entry_date:marker', 'Journal.entry_date', 'Journal.mood_id:marker', 'Journal.mood_id', 'Journal.mood_id.quick_edit', 'Journal.title:marker', 'Journal.title', 'Journal.entry:marker', 'Journal.entry'])
